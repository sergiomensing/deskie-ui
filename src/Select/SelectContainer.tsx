import { PropsWithChildren } from 'react';
import styled, { css } from 'styled-components';
import { KeyboardArrowDown } from '@styled-icons/material-outlined';

export type SelectStatus = 'default' | 'success' | 'warning' | 'error';

export type SelectContainerProps = {
  status?: SelectStatus;
  icon?: React.ReactNode;
  disabled?: boolean;
  className?: string;
};
const StyledSelectContainer = styled.div<{ icon: boolean; disabled: boolean; status: SelectStatus }>`
  display: flex;
  align-items: center;
  position: relative;

  select,
  .select {
    width: 100%;
    display: flex;
    height: 3.5rem;
    padding: 0 0.75rem;
    font: inherit;
    border: 1px solid
      ${(props) => (props.status === 'default' ? props.theme.colors.inputBorder : props.theme.colors[props.status])};
    border-radius: 0.5rem;
    box-shadow: none;
    background: ${(props) => props.theme.colors.inputBackground};
    -webkit-appearance: none;
    color: inherit;
    cursor: pointer;
    position: relative;
    z-index: 1;

    ${(props) => props.icon && `padding-left: 2.25rem;`};
    padding-right: 2.25rem;

    ${(props) =>
      props.disabled &&
      css`
        cursor: not-allowed;
        border: 1px solid ${props.theme.colors.border};
        color: ${props.theme.colors.disabledText};
      `}

    &:focus {
      border-color: ${(props) => props.theme.colors.primary};
      box-shadow: 0px 0px 0px 1px ${(props) => props.theme.colors.primary};
      outline: 0;
    }
  }

  .select-icon {
    position: absolute;
    z-index: 2;
    width: 1.25rem;
    left: 0.75rem;

    svg {
      width: 100%;
      height: auto;
    }
  }

  .select-arrow {
    left: initial;
    right: 0.75rem;
  }
`;

function SelectContainer({
  icon,
  disabled = false,
  children,
  status = 'default',
  className = '',
}: PropsWithChildren<SelectContainerProps>) {
  return (
    <StyledSelectContainer
      disabled={disabled}
      status={status}
      icon={Boolean(icon)}
      className={`select-container ${className}`}
    >
      {icon && <span className="select-icon">{icon}</span>}
      {children}
      <span className="select-icon select-arrow">
        <KeyboardArrowDown />
      </span>
    </StyledSelectContainer>
  );
}

export default SelectContainer;
