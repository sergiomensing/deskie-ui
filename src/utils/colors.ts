export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
