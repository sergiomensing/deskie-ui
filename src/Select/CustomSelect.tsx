import { useState, useCallback, useRef, useEffect, ChangeEventHandler, KeyboardEventHandler } from 'react';
import SelectContainer, { SelectContainerProps } from './SelectContainer';
import styled from 'styled-components';
import { useClickAway } from '../useClickAway';

export type CustomSelectProps = SelectContainerProps & {
  value: string | undefined;
  onChange: (value: string | undefined) => void;
  options: CustomSelectOption[];
  multiple?: boolean;
  className?: string;
  placeholder?: string;
  searchable?: boolean;
  noOptionsLabel?: string;
};

export type CustomSelectOption = {
  label?: string;
  value: string;
};

const StyledCustomSelect = styled.div<{ isActive: boolean }>`
  position: relative;
  width: 100%;

  .select-options {
    position: absolute;
    top: calc(100% + 0.25rem);
    left: 0;
    width: 100%;
    border: 1px solid ${(props) => props.theme.colors.inputBorder};
    background: ${(props) => props.theme.colors.inputBackground};
    box-shadow: ${(props) => props.theme.elevation[1]};
    border-radius: 0.5rem;
    padding: 0.5rem 0;
  }

  .select-option {
    padding: 0.75rem 0.75rem;
    cursor: pointer;

    &.selected {
      background: ${(props) => props.theme.colors.accent2};
    }

    &.current {
      background: ${(props) => props.theme.colors.primary};
      color: ${(props) => props.theme.colors.background};
    }
  }

  .select {
    display: flex;
    align-items: center;
  }

  .select-placeholder {
    color: ${(props) => props.theme.colors.accent8};
  }

  .select-search {
    display: block;
    width: calc(100% - 1.5rem);
    outline: none;
    font: inherit;
    transform: translateX(0px) translateY(-1px);
    position: absolute;
    height: calc(2.5rem);
    border: 0;
    opacity: ${(props) => (props.isActive ? 1 : 0)};
    pointer-events: ${(props) => (props.isActive ? 'initial' : 'none')};
    background: ${(props) => props.theme.colors.background};

    &.select-search-empty {
      background: transparent;
    }
  }

  .select-no-options {
    color: ${(props) => props.theme.colors.accent8};
    padding: 0.75rem 0.75rem;
  }
`;

function CustomSelect({
  options,
  onChange,
  placeholder,
  searchable,
  value,
  disabled,
  noOptionsLabel = 'No options found',
  className,
  icon,
  status,
}: CustomSelectProps) {
  const _value = value ? value.toString() : undefined;
  const [active, setActive] = useState<boolean>(false);
  const [searchValue, setSearchValue] = useState<string>('');
  const searchInputRef = useRef<HTMLInputElement | null>(null);
  const selectRef = useRef<HTMLDivElement | null>(null);
  const [selected, setSelected] = useState<string | undefined>(undefined);

  const filteredOptions = searchable
    ? options.filter((option) => option.value.toString().includes(searchValue))
    : options;

  const _selected =
    searchValue.length > 0 && filteredOptions.length > 0
      ? filteredOptions[0].value
      : selected
      ? selected
      : value
      ? value
      : undefined;

  const handleChange = useCallback(
    (option: CustomSelectOption) => {
      setActive(false);
      setSearchValue('');
      setSelected(undefined);
      onChange(option.value);
    },
    [options],
  );

  const handleSetActive = () => {
    if (!disabled) {
      setActive(searchable ? true : !active);
      if (searchable && searchInputRef.current) {
        searchInputRef.current.focus();
      }
    }
  };

  const handleSearch: ChangeEventHandler<HTMLInputElement> = (event) => {
    setSelected(undefined);
    setSearchValue(event.target.value);
  };

  const handleKeyUp = (event: KeyboardEvent) => {
    if (event.key === 'ArrowDown') {
      event.preventDefault();
      const index = options.findIndex((o) => o.value === _selected);
      if (options[index + 1]) {
        setSelected(options[index + 1].value);
      }
    }

    if (event.key === 'Enter') {
      event.preventDefault();
      const option = options.find((o) => o.value === _selected);
      if (option) {
        handleChange(option);
      }
    }

    if (event.key === 'ArrowUp') {
      event.preventDefault();
      const index = options.findIndex((o) => o.value === _selected);
      if (options[index - 1]) {
        setSelected(options[index - 1].value);
      }
    }
  };

  useClickAway(selectRef, () => {
    setActive(false);
    setSearchValue('');
    setSelected(undefined);
  });

  useEffect(() => {
    if (active) {
      document.addEventListener('keyup', handleKeyUp);

      return () => {
        document.removeEventListener('keyup', handleKeyUp);
      };
    }
  }, [active, selected]);

  return (
    <SelectContainer className={className} icon={icon} disabled={disabled} status={status}>
      <StyledCustomSelect className="custom-select" ref={selectRef} isActive={active}>
        <div className="select" onClick={handleSetActive}>
          {searchable && (
            <input
              className={`select-search ${searchValue.length === 0 ? 'select-search-empty' : ''}`}
              value={searchValue}
              onChange={handleSearch}
              ref={searchInputRef}
            />
          )}
          {_value ? (
            <span className="select-value">{_value}</span>
          ) : (
            <span className="select-placeholder">{placeholder}</span>
          )}
        </div>
        <div
          className="select-options"
          style={{
            opacity: active ? 1 : 0,
            pointerEvents: active ? 'initial' : 'none',
            // transform: active ? 'translateY(0px)' : 'translateY(-0.5rem)',
          }}
        >
          {filteredOptions.length === 0 && <div className="select-no-options">{noOptionsLabel}</div>}
          {filteredOptions.map((option) => (
            <div
              onMouseEnter={() => setSelected(option.value)}
              className={`select-option ${option.value === _selected ? 'selected' : ''} ${
                value === option.value ? 'current' : ''
              } `}
              onClick={() => handleChange(option)}
              key={option.value}
            >
              {option.label || option.value}
            </div>
          ))}
        </div>
      </StyledCustomSelect>
    </SelectContainer>
  );
}

export default CustomSelect;
