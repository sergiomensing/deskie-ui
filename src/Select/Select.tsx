import { ChangeEvent, forwardRef, PropsWithChildren, SelectHTMLAttributes, useImperativeHandle, useRef } from 'react';
import SelectContainer, { SelectContainerProps } from './SelectContainer';

export type SelectProps = SelectContainerProps &
  Omit<SelectHTMLAttributes<unknown>, 'onChange'> & {
    onChange?: (ev: ChangeEvent<HTMLSelectElement>) => void;
  };

const Select = forwardRef<HTMLSelectElement, PropsWithChildren<SelectProps>>(
  (
    { disabled = false, icon, status = 'default', children, className, ...rest },
    ref: React.Ref<HTMLSelectElement | null>,
  ) => {
    const selectRef = useRef<HTMLSelectElement>(null);
    useImperativeHandle(ref, () => selectRef.current);

    return (
      <SelectContainer className={className} icon={icon} disabled={disabled} status={status}>
        <select ref={selectRef} disabled={disabled} {...rest}>
          {children}
        </select>
      </SelectContainer>
    );
  },
);

export default Select;
