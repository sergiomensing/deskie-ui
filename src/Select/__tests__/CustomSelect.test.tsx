import React, { useState } from 'react';
import Select, { CustomSelectProps } from '../CustomSelect';
import { fireEvent, render } from '../../../tests/utils';

const options = [{ value: 'option1' }, { value: 'option2' }, { value: 'option3' }];

const Wrapper = ({
  value,
  ...rest
}: Omit<CustomSelectProps, 'value' | 'onChange' | 'options'> & { value?: string }) => {
  const [selected, setSelected] = useState<string | undefined>(() => value);
  return <Select value={selected} onChange={setSelected} options={options} {...rest} />;
};

describe('Custom Select', () => {
  it('should render', () => {
    const { getByText, container } = render(<Wrapper placeholder="test-value" />);
    expect(getByText('test-value')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should show options', () => {
    const { getByText, container } = render(<Wrapper placeholder="test-value" />);
    expect(window.getComputedStyle(container.querySelector('.select-options')).opacity).toBe('0');
    fireEvent.click(getByText('test-value'));
    expect(window.getComputedStyle(container.querySelector('.select-options')).opacity).toBe('1');
  });

  it('should show value', () => {
    const { getByText } = render(<Wrapper value="test-value" />);
    expect(getByText('test-value')).toBeInTheDocument();
  });

  it('should change value', () => {
    const { getByText, container } = render(<Wrapper value="test-value" />);
    fireEvent.click(getByText('test-value'));
    fireEvent.click(getByText('option1'));
    expect(container.querySelector('.select-value').textContent).toBe('option1');
  });

  it('should not display options if disabled', () => {
    const { getByText, container } = render(<Wrapper value="test-value" disabled />);
    fireEvent.click(getByText('test-value'));
    expect(window.getComputedStyle(container.querySelector('.select-options')).opacity).toBe('0');
  });
});
