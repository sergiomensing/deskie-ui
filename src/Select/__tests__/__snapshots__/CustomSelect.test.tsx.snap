// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Select should render 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c0 select,
.c0 .select {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3.5rem;
  padding: 0 0.75rem;
  font: inherit;
  border: 1px solid #9395A1;
  border-radius: 0.5rem;
  box-shadow: none;
  background: #FFFFFF;
  -webkit-appearance: none;
  color: inherit;
  cursor: pointer;
  position: relative;
  z-index: 1;
  padding-right: 2.25rem;
}

.c0 select:focus,
.c0 .select:focus {
  border-color: #6C5DD3;
  box-shadow: 0px 0px 0px 1px #6C5DD3;
  outline: 0;
}

.c0 .select-icon {
  position: absolute;
  z-index: 2;
  width: 1.25rem;
  left: 0.75rem;
}

.c0 .select-icon svg {
  width: 100%;
  height: auto;
}

.c0 .select-arrow {
  left: initial;
  right: 0.75rem;
}

.c1 {
  position: relative;
  width: 100%;
}

.c1 .select-options {
  position: absolute;
  top: calc(100% + 0.25rem);
  left: 0;
  width: 100%;
  border: 1px solid #9395A1;
  background: #FFFFFF;
  box-shadow: 0px 3px 6px rgba(0,0,0,0.04),0px 1px 2px rgba(0,0,0,0.02);
  border-radius: 0.5rem;
  padding: 0.5rem 0;
}

.c1 .select-option {
  padding: 0.75rem 0.75rem;
  cursor: pointer;
}

.c1 .select-option.selected {
  background: #EEEEF8;
}

.c1 .select-option.current {
  background: #6C5DD3;
  color: #FFFFFF;
}

.c1 .select {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 .select-placeholder {
  color: #9395A1;
}

.c1 .select-search {
  display: block;
  width: calc(100% - 1.5rem);
  outline: none;
  font: inherit;
  -webkit-transform: translateX(0px) translateY(-1px);
  -ms-transform: translateX(0px) translateY(-1px);
  transform: translateX(0px) translateY(-1px);
  position: absolute;
  height: calc(2.5rem);
  border: 0;
  opacity: 0;
  pointer-events: none;
  background: #FFFFFF;
}

.c1 .select-search.select-search-empty {
  background: transparent;
}

.c1 .select-no-options {
  color: #9395A1;
  padding: 0.75rem 0.75rem;
}

<div>
  <div
    class="c0 select-container "
  >
    <div
      class="c1 custom-select"
    >
      <div
        class="select"
      >
        <span
          class="select-placeholder"
        >
          test-value
        </span>
      </div>
      <div
        class="select-options"
        style="opacity: 0; pointer-events: none;"
      >
        <div
          class="select-option   "
        >
          option1
        </div>
        <div
          class="select-option   "
        >
          option2
        </div>
        <div
          class="select-option   "
        >
          option3
        </div>
      </div>
    </div>
    <span
      class="select-icon select-arrow"
    >
      <svg
        aria-hidden="true"
        class="c2"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
        />
      </svg>
    </span>
  </div>
</div>
`;
